runtime/near-vm-runner/RUNTIMES.md

```diff
@@ -0,0 +1,176 @@
+would be to directly translate all the stack operations to native code as well. Targets running
+this code, on the other hand, are most likely an implementation of register machine where stack
+operations are not necessarily all that efficient.
+
+In order to improve the quality of the generated machine code, `wasmer-singlepass` maintains some
+state between the individual Wasm instructions. This state allows `wasmer-singlepass` to avoid some
+stack operations, pick more efficient instructions and sometimes avoid some operations such as NaN
+canonicalization altogether.
+
+Presence of this state comes at a cost of testability of implementation. The behaviour of the
+codegen can depend on prior Wasm instructions. Because of this there's a combinatorial explosion of
+the test cases that need to be considered and verified before confidence in the results can be
+established. Relatedly, at times it may also be difficult to validate changes made to the
+implementation of the `wasmer-singlepass` codegen. The global state is definitely a source of
+potential spooky action at a distance problems where changing code generation of a specific
+instruction affects correctness and behaviour of another one.
+
+The `wasmer-singlepass` backend currently only supports `x86_64` with `AVX` on Linux and, since
+very recently, Windows.
+
+## Cranelift codegen
+
+Cranelift is a code generator written in Rust. It aims to be a fast code generator that also
+produces machine code of a reasonably high quality. Originally it was designed for use in Firefox
+for compiling arbitrary Wasm binaries presented to the browser by websites. Since then Cranelift
+can be used for code generation in both Wasmer and Wasmtime.
+
+Cranelift is a multi-stage compiler. It consumes an intermediate representation called CLIF which
+is later lowered to the VCode IR on which operations such as instruction selection and register
+allocation are applied to produce the machine code. Refer to [this blog post][clift-cg-primer] for
+an in-depth explanation of the Cranelift's architecture.
+
+[clift-cg-primer]: https://blog.benj.me/2021/02/17/cranelift-codegen-primer/
+
+A Wasm codegen implementation in a browser shares many of the requirements with those of the NEAR
+validator. Cranelift gets high scores for its Correctness, Reliability and Security properties.
+It supports targeting the Arm, s390x and x86_64 architectures as well as a variety of
+operating systems, netting it a high score on the platform support metric as well.
+
+Upon a brief investigation in December of 2021 generating machine code with Cranelift should
+broadly be `O(n)` over input size with the exception of register allocation, which due to their
+soon-to-be replaced regalloc algorithm is currently `O(n²)`.
+
+One detail where Cranelift may fall short is in the ability to produce super-optimized machine code
+sequences for hot operations such as gas counting.
```

Comment:
I feel something's missing here... Especially, my main worry with cranelift is that it'd be hard to estimate costs for generated code. Basically, with `singlepass` we somewhat confidente that the compiler isn't being overly smart. With cratelift, it feels like it could happen that we are benchmarking happy case.